File: spring-data-requery-test/src/main/java/org/springframework/data/requery/domain/upsert/AbstractUpsertPlace.java
package org.springframework.data.requery.domain.upsert;

import io.requery.*;
import io.requery.query.MutableResult;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;
import org.springframework.data.requery.domain.AbstractPersistable;
import org.springframework.data.requery.domain.ToStringBuilder;

import java.util.Objects;

/**
 * @author Diego on 2018. 6. 10..
 */
@Getter
@Entity(name = "UpsertPlace")
@Table(name = "upsert_place")
public abstract class AbstractUpsertPlace extends AbstractPersistable<String> {

    @Key
    protected String id;
    protected String name;

    @OneToMany(mappedBy = "place", cascade = { CascadeAction.SAVE })
    protected MutableResult<AbstractUpsertEvent> events;

    @Override
    public int hashCode() {
        return Objects.hash(name);
    }

    @Transient
    @Override
    protected @NotNull ToStringBuilder buildStringHelper() {
        return super.buildStringHelper()
            .add("name", name);
    }
}

File: src/test/kotlin/com/coupang/springframework/data/requery/domain/upsert/AbstractUpsertPlace.kt
package com.coupang.springframework.data.requery.domain.upsert

import com.coupang.kotlinx.core.hashOf
import com.coupang.kotlinx.objectx.ToStringBuilder
import com.coupang.springframework.data.requery.domain.AbstractPersistable
import io.requery.*


@Entity
abstract class AbstractUpsertPlace: AbstractPersistable<String>() {

    @get:Key
    abstract override var id: String?

    @get:Column
    abstract var name: String

    @get:OneToMany(mappedBy = "place", cascade = [CascadeAction.SAVE])
    abstract val events: MutableSet<AbstractUpsertEvent>

    fun addEvent(event: AbstractUpsertEvent) {
        this.events.add(event)
        event.place = this
    }

    override fun hashCode(): Int {
        return hashOf(name)
    }

    @Transient
    override fun buildStringHelper(): ToStringBuilder {
        return super.buildStringHelper()
            .add("name", name)
    }
}